File: 模板/树/树状数组.cpp
#include <vector>

using namespace std;

template <typename T>
class BinaryIndexTree { // tree[i]包含[i - lowbit(i) + 1, i]信息
public:
    vector<T> tree{};
    int size{};

    explicit BinaryIndexedTree(int size): size(size) {
        tree.resize(size + 1);
    }

    void change(int loc, T delta) {
        while (loc <= size) {
            tree[loc] += delta;
            loc += loc & -loc;
        }
    }

    T query(int loc) {
        T ans{};
        while (loc) {
            ans += tree[loc];
            loc ^= loc & -loc;
        }
        return ans;
    }
};

File: 模板/python/树/树状数组.py
class BinaryIndexTree:  # tree[i]包含[i - lowbit(i) + 1, i]信息
    __slots__ = ["length", "tree"]

    def __init__(self, length):
        self.length = length
        self.tree = [0] * (length + 1)

    def change(self, loc, delta):
        while loc <= self.length:
            self.tree[loc] += delta
            loc += loc & -loc

    def query(self, loc):
        ans = 0
        while loc:
            ans += self.tree[loc]
            loc ^= loc & -loc
        return ans